File: python/adapters/AzureOpenAIAdapter.py
"""
Azure OpenAI Adapter for Python
Implements ILLMProvider interface for Azure OpenAI
"""

import asyncio
import logging
from datetime import datetime
from typing import Dict, List, Optional, Any
from openai import AsyncAzureOpenAI

from core.interfaces.ILLMProvider import ILLMProvider, LLMResponse, LLMMessage

class AzureOpenAIAdapter(ILLMProvider):
    """Azure OpenAI implementation of ILLMProvider"""
    
    def __init__(self, container):
        self.container = container
        self.config = container.resolve('ConfigurationManager')
        self.logger = logging.getLogger(__name__)
        
        # Get Azure OpenAI configuration
        self.azure_config = self.config.get_section('llm.azureOpenAI')
        
        self.client = None
        self.is_initialized = False
        
        # Token pricing (approximate rates per 1M tokens)
        self.pricing = {
            'gpt-4': {'input': 30, 'output': 60},
            'gpt-4-turbo': {'input': 10, 'output': 30},
            'gpt-35-turbo': {'input': 0.5, 'output': 1.5}
        }

    async def initialize(self) -> bool:
        """Initialize Azure OpenAI client"""
        try:
            endpoint = self.azure_config.get('endpoint')
            api_key = self.azure_config.get('apiKey')
            deployment = self.azure_config.get('deployment')
            
            self.logger.info(f"🔧 Azure OpenAI config: endpoint={endpoint}, deployment={deployment}, apiKey={'set' if api_key else 'MISSING'}")
            
            if not all([endpoint, api_key, deployment]):
                missing = []
                if not endpoint: missing.append('endpoint')
                if not api_key: missing.append('apiKey')
                if not deployment: missing.append('deployment')
                raise ValueError(f"Missing required Azure OpenAI configuration: {', '.join(missing)}")

            self.client = AsyncAzureOpenAI(
                api_key=api_key,
                api_version=self.azure_config.get('apiVersion', '2024-02-01'),
                azure_endpoint=endpoint
            )
            
            self.is_initialized = True
            self.logger.info(f"✅ Azure OpenAI initialized: {deployment}")
            return True
            
        except Exception as e:
            self.logger.error(f"❌ Failed to initialize Azure OpenAI: {e}")
            return False

    async def _test_connection(self):
        """Test the Azure OpenAI connection"""
        try:
            response = await self.client.chat.completions.create(
                model=self.azure_config.get('deployment'),
                messages=[{"role": "user", "content": "Test"}],
                max_tokens=5
            )
            
            if not response.choices:
                raise ValueError("Invalid response from Azure OpenAI")
                
        except Exception as e:
            raise ConnectionError(f"Azure OpenAI connection test failed: {e}")

    async def is_ready(self) -> bool:
        """Check if provider is ready"""
        return self.is_initialized and self.client is not None

    def get_provider_name(self) -> str:
        """Get provider name"""
        return "Azure OpenAI"

    def get_model_name(self) -> str:
        """Get model name"""
        return self.azure_config.get('deployment', 'gpt-4')

    async def generate_response(
        self,
        messages: List[LLMMessage],
        max_tokens: Optional[int] = None,
        temperature: Optional[float] = None,
        **kwargs
    ) -> LLMResponse:
        """Generate response from messages"""
        if not await self.is_ready():
            raise RuntimeError("Azure OpenAI provider not ready")

        try:
            # Convert to OpenAI format
            openai_messages = [
                {
                    "role": msg.role,
                    "content": msg.content,
                    **({"name": msg.name} if msg.name else {})
                }
                for msg in messages
            ]

            # Make request
            response = await self.client.chat.completions.create(
                model=self.azure_config.get('deployment'),
                messages=openai_messages,
                max_tokens=max_tokens or self.azure_config.get('maxTokens', 4096),
                temperature=temperature or self.azure_config.get('temperature', 0.7),
                top_p=self.azure_config.get('topP', 0.95),
                **kwargs
            )

            # Format response
            choice = response.choices[0]
            usage = response.usage or {}

            return LLMResponse(
                content=choice.message.content or "",
                usage={
                    "prompt_tokens": usage.prompt_tokens or 0,
                    "completion_tokens": usage.completion_tokens or 0,
                    "total_tokens": usage.total_tokens or 0
                },
                model=response.model or self.get_model_name(),
                provider=self.get_provider_name(),
                finish_reason=choice.finish_reason or "stop",
                timestamp=datetime.utcnow().isoformat()
            )

        except Exception as e:
            self.logger.error(f"❌ Azure OpenAI generation failed: {e}")
            raise

    async def generate_with_tools(
        self,
        messages: List[LLMMessage],
        tools: List[Dict] = None,
        **kwargs
    ) -> LLMResponse:
        """Generate response with tool calling"""
        if not tools:
            return await self.generate_response(messages, **kwargs)

        try:
            # Convert tools to OpenAI format
            openai_tools = [
                {
                    "type": "function",
                    "function": {
                        "name": tool["name"],
                        "description": tool["description"],
                        "parameters": tool["parameters"]
                    }
                }
                for tool in tools
            ]

            # Convert messages
            openai_messages = [
                {"role": msg.role, "content": msg.content}
                for msg in messages
            ]

            response = await self.client.chat.completions.create(
                model=self.azure_config.get('deployment'),
                messages=openai_messages,
                tools=openai_tools,
                tool_choice="auto",
                **kwargs
            )

            choice = response.choices[0]
            usage = response.usage or {}

            # Handle tool calls if present
            content = choice.message.content or ""
            if choice.message.tool_calls:
                # Add tool calls to response
                content += f"\n\nTool calls: {len(choice.message.tool_calls)}"

            return LLMResponse(
                content=content,
                usage={
                    "prompt_tokens": usage.prompt_tokens or 0,
                    "completion_tokens": usage.completion_tokens or 0,
                    "total_tokens": usage.total_tokens or 0
                },
                model=response.model or self.get_model_name(),
                provider=self.get_provider_name(),
                finish_reason=choice.finish_reason or "stop",
                timestamp=datetime.utcnow().isoformat()
            )

        except Exception as e:
            self.logger.error(f"❌ Azure OpenAI tool generation failed: {e}")
            raise

    def get_autogen_config(self) -> Dict[str, Any]:
        """Get AutoGen compatible configuration"""
        return {
            "config_list": [{
                "model": self.azure_config.get('deployment'),
                "api_type": "azure",
                "api_base": self.azure_config.get('endpoint'),
                "api_key": self.azure_config.get('apiKey'),
                "api_version": self.azure_config.get('apiVersion')
            }],
            "temperature": self.azure_config.get('temperature', 0.7),
            "max_tokens": self.azure_config.get('maxTokens', 4096),
            "top_p": self.azure_config.get('topP', 0.95)
        }

    async def estimate_tokens(self, text: str) -> int:
        """Estimate token count (rough approximation)"""
        # Rough estimation: 1 token ≈ 0.75 words
        words = len(text.split())
        return int(words / 0.75)

    async def calculate_cost(self, input_tokens: int, output_tokens: int) -> Dict:
        """Calculate cost for token usage"""
        model_key = self.get_model_name().lower()
        rates = self.pricing.get('gpt-4', self.pricing['gpt-4'])  # Default
        
        for key, value in self.pricing.items():
            if key in model_key:
                rates = value
                break

        input_cost = (input_tokens / 1000000) * rates['input']
        output_cost = (output_tokens / 1000000) * rates['output']
        
        return {
            "input_cost": input_cost,
            "output_cost": output_cost,
            "total_cost": input_cost + output_cost,
            "currency": "USD"
        }

    async def supports_vision(self) -> bool:
        """Check if model supports vision"""
        model = self.get_model_name().lower()
        return 'vision' in model or 'gpt-4' in model

    async def supports_tools(self) -> bool:
        """Check if model supports function calling"""
        return True  # Most Azure OpenAI models support tools

    async def supports_streaming(self) -> bool:
        """Check if provider supports streaming"""
        return True

    async def get_health_status(self) -> Dict:
        """Get provider health status"""
        try:
            if await self.is_ready():
                await self._test_connection()
                status = "healthy"
            else:
                status = "not_ready"
        except Exception as e:
            status = f"unhealthy: {str(e)}"

        return {
            "provider": self.get_provider_name(),
            "model": self.get_model_name(),
            "status": status,
            "ready": await self.is_ready(),
            "endpoint": self.azure_config.get('endpoint', 'unknown'),
            "deployment": self.azure_config.get('deployment', 'unknown')
        }

File: nodejs/src/llm/adapters/AzureOpenAIAdapter.js
/**
 * Azure OpenAI Adapter
 * Implements ILLMProvider interface for Azure OpenAI GPT models
 * Supports GPT-4, GPT-4-turbo, vision, function calling, and embeddings
 */

const { OpenAI } = require('openai');
const ILLMProvider = require('../../core/interfaces/ILLMProvider');
const logger = require('../../utils/logger');

class AzureOpenAIAdapter extends ILLMProvider {
    constructor(config = {}) {
        super();
        this.config = {
            endpoint: config.endpoint || process.env.AZURE_OPENAI_ENDPOINT,
            apiKey: config.apiKey || process.env.AZURE_OPENAI_API_KEY,
            deployment: config.deployment || process.env.AZURE_OPENAI_DEPLOYMENT,
            apiVersion: config.apiVersion || process.env.AZURE_OPENAI_API_VERSION || '2024-02-01',
            maxTokens: config.maxTokens || parseInt(process.env.AZURE_OPENAI_MAX_TOKENS) || 4096,
            temperature: config.temperature || parseFloat(process.env.AZURE_OPENAI_TEMPERATURE) || 0.7,
            topP: config.topP || parseFloat(process.env.AZURE_OPENAI_TOP_P) || 0.95,
            ...config
        };

        this.client = null;
        this.isReady = false;
        this.healthStats = {
            provider: 'Azure OpenAI',
            model: this.config.deployment,
            status: 'initializing',
            lastCheck: null,
            uptime: 0,
            errorCount: 0,
            totalRequests: 0,
            startTime: Date.now()
        };

        // Token pricing (approximate rates per 1M tokens)
        this.pricing = {
            'gpt-4': { input: 30, output: 60 },
            'gpt-4-turbo': { input: 10, output: 30 },
            'gpt-4-32k': { input: 60, output: 120 },
            'gpt-3.5-turbo': { input: 0.5, output: 1.5 },
            'gpt-3.5-turbo-16k': { input: 3, output: 4 }
        };
    }

    async initialize() {
        try {
            if (!this.config.endpoint || !this.config.apiKey || !this.config.deployment) {
                throw new Error('Missing required Azure OpenAI configuration');
            }

            this.client = new OpenAI({
                apiKey: this.config.apiKey,
                baseURL: `${this.config.endpoint}/openai/deployments/${this.config.deployment}`,
                defaultQuery: { 'api-version': this.config.apiVersion },
                defaultHeaders: {
                    'api-key': this.config.apiKey,
                }
            });

            await this.testConnection();
            this.isReady = true;
            this.healthStats.status = 'ready';
            
            logger.info(`✅ Azure OpenAI initialized: ${this.config.deployment}`);
        } catch (error) {
            this.healthStats.status = 'error';
            this.healthStats.errorCount++;
            logger.error('❌ Failed to initialize Azure OpenAI:', error);
            throw error;
        }
    }

    async testConnection() {
        try {
            const response = await this.client.chat.completions.create({
                messages: [{ role: 'user', content: 'Test connection' }],
                max_tokens: 10,
                temperature: 0
            });

            if (response.choices && response.choices.length > 0) {
                this.healthStats.lastCheck = new Date();
                return 'Connection successful';
            } else {
                throw new Error('Invalid response from Azure OpenAI');
            }
        } catch (error) {
            this.healthStats.errorCount++;
            throw error;
        }
    }

    async isReady() {
        return this.isReady && this.client !== null;
    }

    getModelName() {
        return this.config.deployment;
    }

    getProviderName() {
        return 'Azure OpenAI';
    }

    getMaxTokens() {
        return this.config.maxTokens;
    }

    getContextWindow() {
        const model = this.config.deployment.toLowerCase();
        if (model.includes('gpt-4-32k')) return 32768;
        if (model.includes('gpt-4')) return 8192;
        if (model.includes('16k')) return 16384;
        return 4096;
    }

    async estimateTokens(text) {
        // Rough estimation: 1 token ≈ 0.75 words
        const words = text.split(/\s+/).length;
        return Math.ceil(words / 0.75);
    }

    async calculateCost(inputTokens, outputTokens) {
        const modelKey = this.config.deployment.toLowerCase();
        let rates = this.pricing['gpt-4']; // Default
        
        for (const [key, value] of Object.entries(this.pricing)) {
            if (modelKey.includes(key)) {
                rates = value;
                break;
            }
        }

        const inputCost = (inputTokens / 1000000) * rates.input;
        const outputCost = (outputTokens / 1000000) * rates.output;
        
        return {
            inputCost,
            outputCost,
            totalCost: inputCost + outputCost,
            currency: 'USD'
        };
    }

    async generateResponse(messages, options = {}) {
        try {
            this.healthStats.totalRequests++;
            
            const requestOptions = {
                messages: this.formatMessages(messages),
                max_tokens: options.maxTokens || this.config.maxTokens,
                temperature: options.temperature ?? this.config.temperature,
                top_p: options.topP ?? this.config.topP,
                frequency_penalty: options.frequencyPenalty || 0,
                presence_penalty: options.presencePenalty || 0,
                stop: options.stop || null,
                ...options.extra
            };

            const response = await this.client.chat.completions.create(requestOptions);
            
            const result = {
                content: response.choices[0]?.message?.content || '',
                finishReason: response.choices[0]?.finish_reason || 'unknown',
                usage: {
                    promptTokens: response.usage?.prompt_tokens || 0,
                    completionTokens: response.usage?.completion_tokens || 0,
                    totalTokens: response.usage?.total_tokens || 0
                },
                model: response.model || this.config.deployment,
                provider: 'Azure OpenAI',
                timestamp: new Date().toISOString()
            };

            // Calculate costs
            if (result.usage.totalTokens > 0) {
                result.cost = await this.calculateCost(
                    result.usage.promptTokens, 
                    result.usage.completionTokens
                );
            }

            this.healthStats.uptime = Date.now() - this.healthStats.startTime;
            return result;

        } catch (error) {
            this.healthStats.errorCount++;
            return this.handleAPIError(error);
        }
    }

    async generateStreamResponse(messages, options = {}, onChunk) {
        try {
            this.healthStats.totalRequests++;
            
            const requestOptions = {
                messages: this.formatMessages(messages),
                max_tokens: options.maxTokens || this.config.maxTokens,
                temperature: options.temperature ?? this.config.temperature,
                top_p: options.topP ?? this.config.topP,
                stream: true,
                ...options.extra
            };

            const stream = await this.client.chat.completions.create(requestOptions);
            
            let fullContent = '';
            let usage = { promptTokens: 0, completionTokens: 0, totalTokens: 0 };

            for await (const chunk of stream) {
                const delta = chunk.choices[0]?.delta?.content || '';
                if (delta) {
                    fullContent += delta;
                    if (onChunk) {
                        await onChunk({
                            content: delta,
                            fullContent,
                            isComplete: false
                        });
                    }
                }

                if (chunk.usage) {
                    usage = {
                        promptTokens: chunk.usage.prompt_tokens || 0,
                        completionTokens: chunk.usage.completion_tokens || 0,
                        totalTokens: chunk.usage.total_tokens || 0
                    };
                }
            }

            const result = {
                content: fullContent,
                finishReason: 'stop',
                usage,
                model: this.config.deployment,
                provider: 'Azure OpenAI',
                timestamp: new Date().toISOString()
            };

            if (onChunk) {
                await onChunk({
                    content: '',
                    fullContent,
                    isComplete: true,
                    usage: result.usage
                });
            }

            return result;

        } catch (error) {
            this.healthStats.errorCount++;
            throw this.handleAPIError(error);
        }
    }

    async generateEmbedding(text) {
        try {
            // Note: You need a separate embedding deployment
            const response = await this.client.embeddings.create({
                input: text,
                model: 'text-embedding-ada-002' // Or your embedding deployment name
            });

            return {
                embedding: response.data[0].embedding,
                usage: response.usage,
                model: response.model,
                provider: 'Azure OpenAI'
            };

        } catch (error) {
            logger.warn('Azure OpenAI embedding not available:', error.message);
            return null;
        }
    }

    async classifyIntent(message, classes) {
        const prompt = `Classify the following message into one of these categories: ${classes.join(', ')}.
        
Message: "${message}"

Respond with only the category name.`;

        const response = await this.generateResponse([
            { role: 'user', content: prompt }
        ], { maxTokens: 50, temperature: 0 });

        const classification = response.content.trim();
        return {
            intent: classification,
            confidence: classes.includes(classification) ? 0.9 : 0.1
        };
    }

    async summarizeConversation(messages, maxLength = 200) {
        const conversation = messages.map(m => `${m.role}: ${m.content}`).join('\n');
        const prompt = `Summarize this conversation in ${maxLength} characters or less:

${conversation}

Summary:`;

        const response = await this.generateResponse([
            { role: 'user', content: prompt }
        ], { maxTokens: 100, temperature: 0.3 });

        return response.content.trim();
    }

    async generateWithTools(messages, tools = [], options = {}) {
        try {
            const requestOptions = {
                messages: this.formatMessages(messages),
                max_tokens: options.maxTokens || this.config.maxTokens,
                temperature: options.temperature ?? this.config.temperature,
                tools: tools.map(tool => ({
                    type: 'function',
                    function: {
                        name: tool.name,
                        description: tool.description,
                        parameters: tool.parameters
                    }
                })),
                tool_choice: options.toolChoice || 'auto',
                ...options.extra
            };

            const response = await this.client.chat.completions.create(requestOptions);
            
            const result = {
                content: response.choices[0]?.message?.content || '',
                toolCalls: response.choices[0]?.message?.tool_calls || [],
                finishReason: response.choices[0]?.finish_reason || 'unknown',
                usage: {
                    promptTokens: response.usage?.prompt_tokens || 0,
                    completionTokens: response.usage?.completion_tokens || 0,
                    totalTokens: response.usage?.total_tokens || 0
                },
                model: response.model || this.config.deployment,
                provider: 'Azure OpenAI'
            };

            return result;

        } catch (error) {
            this.healthStats.errorCount++;
            return this.handleAPIError(error);
        }
    }

    supportsTools() {
        return true;
    }

    async analyzeImage(imageData, prompt = "Describe this image") {
        try {
            // Check if model supports vision
            if (!this.config.deployment.includes('vision') && !this.config.deployment.includes('gpt-4')) {
                throw new Error('Vision not supported by this model');
            }

            const messages = [{
                role: 'user',
                content: [
                    { type: 'text', text: prompt },
                    { 
                        type: 'image_url', 
                        image_url: { 
                            url: `data:image/jpeg;base64,${imageData}`,
                            detail: 'auto'
                        } 
                    }
                ]
            }];

            const response = await this.generateResponse(messages);
            return response;

        } catch (error) {
            logger.warn('Azure OpenAI vision not available:', error.message);
            throw error;
        }
    }

    supportsVision() {
        return this.config.deployment.includes('vision') || this.config.deployment.includes('gpt-4');
    }

    async speechToText(audioData) {
        try {
            const response = await this.client.audio.transcriptions.create({
                file: audioData,
                model: 'whisper-1',
                response_format: 'text'
            });

            return {
                text: response,
                provider: 'Azure OpenAI',
                model: 'whisper-1'
            };

        } catch (error) {
            logger.warn('Azure OpenAI speech-to-text not available:', error.message);
            throw error;
        }
    }

    async textToSpeech(text, voice = 'alloy') {
        try {
            const response = await this.client.audio.speech.create({
                model: 'tts-1',
                voice: voice,
                input: text,
                response_format: 'mp3'
            });

            return response;

        } catch (error) {
            logger.warn('Azure OpenAI text-to-speech not available:', error.message);
            throw error;
        }
    }

    supportsSpeech() {
        return true; // Azure OpenAI supports both speech-to-text and text-to-speech
    }

    getAutoGenConfig() {
        return {
            config_list: [{
                model: this.config.deployment,
                api_type: 'azure',
                api_base: this.config.endpoint,
                api_key: this.config.apiKey,
                api_version: this.config.apiVersion
            }],
            temperature: this.config.temperature,
            max_tokens: this.config.maxTokens,
            top_p: this.config.topP
        };
    }

    async handleAPIError(error, retryCount = 0) {
        const maxRetries = 3;
        const baseDelay = 1000;

        // Specific error handling
        if (error.status === 429 && retryCount < maxRetries) {
            // Rate limit - exponential backoff
            const delay = baseDelay * Math.pow(2, retryCount);
            logger.warn(`Rate limited, retrying in ${delay}ms...`);
            
            await new Promise(resolve => setTimeout(resolve, delay));
            return this.handleAPIError(error, retryCount + 1);
        }

        if (error.status === 503 && retryCount < maxRetries) {
            // Service unavailable - retry
            const delay = baseDelay * (retryCount + 1);
            logger.warn(`Service unavailable, retrying in ${delay}ms...`);
            
            await new Promise(resolve => setTimeout(resolve, delay));
            return this.handleAPIError(error, retryCount + 1);
        }

        // Log error details
        logger.error('Azure OpenAI API Error:', {
            status: error.status,
            message: error.message,
            type: error.type,
            retryCount
        });

        throw new Error(`Azure OpenAI Error: ${error.message}`);
    }

    async checkRateLimit() {
        // Implement custom rate limiting if needed
        return true;
    }

    async switchModel(modelName) {
        this.config.deployment = modelName;
        this.healthStats.model = modelName;
        
        // Update client with new deployment
        this.client = new OpenAI({
            apiKey: this.config.apiKey,
            baseURL: `${this.config.endpoint}/openai/deployments/${modelName}`,
            defaultQuery: { 'api-version': this.config.apiVersion },
            defaultHeaders: {
                'api-key': this.config.apiKey,
            }
        });

        logger.info(`✅ Switched to model: ${modelName}`);
    }

    getAvailableModels() {
        // Return commonly available Azure OpenAI models
        return [
            'gpt-4',
            'gpt-4-turbo',
            'gpt-4-32k',
            'gpt-35-turbo',
            'gpt-35-turbo-16k'
        ];
    }

    getHealthStatus() {
        this.healthStats.uptime = Date.now() - this.healthStats.startTime;
        return { ...this.healthStats };
    }

    async performHealthCheck() {
        try {
            await this.testConnection();
            this.healthStats.status = 'healthy';
            this.healthStats.lastCheck = new Date();
        } catch (error) {
            this.healthStats.status = 'unhealthy';
            this.healthStats.errorCount++;
            throw error;
        }
    }

    formatMessages(messages) {
        return messages.map(msg => ({
            role: msg.role,
            content: msg.content,
            name: msg.name || undefined
        }));
    }
}

module.exports = AzureOpenAIAdapter;
